src/io/windows.zig

```diff
@@ -0,0 +1,1132 @@
+
+    fn flush(self: *IO, mode: FlushMode) !void {
+        if (self.completed.peek() == null) {
+            // Compute how long to poll by flushing timeout completions.
+            // NOTE: this may push to completed queue
+            var timeout_ms: ?os.windows.DWORD = null;
+            if (self.flush_timeouts()) |expires_ns| {
+                // 0ns expires should have been completed not returned
+                assert(expires_ns != 0); 
+                // Round up sub-millisecond expire times to the next millisecond
+                const expires_ms = (expires_ns + (std.time.ns_per_ms / 2)) / std.time.ns_per_ms;
+                // Saturating cast to DWORD milliseconds
+                const expires = std.math.cast(os.windows.DWORD, expires_ms) catch std.math.maxInt(os.windows.DWORD);
+                // max DWORD is reserved for INFINITE so cap the cast at max - 1
+                timeout_ms = if (expires == os.windows.INFINITE) expires - 1 else expires;
+            }
+            
+            // Poll for IO iff theres IO pending and flush_timeouts() found no ready completions
+            if (self.io_pending > 0 and self.completed.peek() == null) {
+                // In blocking mode, we're always waiting at least until the timeout by run_for_ns.
+                // In non-blocking mode, we shouldn't wait at all.
+                const io_timeout = switch (mode) {
+                    .blocking => timeout_ms orelse @panic("IO.flush blocking unbounded"),
+                    .non_blocking => 0,
+                };
+
+                var events: [64]os.windows.OVERLAPPED_ENTRY = undefined;
+                const num_events = os.windows.GetQueuedCompletionStatusEx(
+                    self.iocp,
+                    &events,
+                    io_timeout,
+                    false, // non-alertable wait
+                ) catch |err| switch (err) {
+                    error.Timeout => 0,
+                    error.Aborted => unreachable,
+                    else => |e| return e,
+                };
+                
+                assert(self.io_pending >= num_events);
+                self.io_pending -= num_events;
+
+                for (events[0..num_events]) |event| {
+                    const raw_overlapped = event.lpOverlapped;
+                    const overlapped = @fieldParentPtr(Completion.Overlapped, "raw", raw_overlapped);
+                    const completion = overlapped.completion;
+                    completion.next = null;
+                    self.completed.push(completion);
+                }
+            }
+        }
+
+        // Dequeue and invoke all the completions currently ready.
+        // Must read all `completions` before invoking the callbacks
+        // as the callbacks could potentially submit more completions. 
+        var completed = self.completed;
+        self.completed = .{};
+        while (completed.pop()) |completion| {
+            (completion.callback)(Completion.Context{
```

Comment:
Are these parentheses really necessary?
```suggestion
            completion.callback(Completion.Context{
```

Reply:
Whoops, habits from writing other langs leaking in. (It disambiguates static method call from function pointer field call).